app/gui2/src/stores/visualization.ts

```diff
@@ -291,21 +437,151 @@ export const useVisualizationStore = defineStore('visualization', () => {
     return Promise.allSettled(promises)
   }
 
-  // NOTE: Because visualization scripts are cached, they are not guaranteed to be up to date.
-  async function get(meta: VisualizationIdentifier) {
-    if (meta.module.kind !== 'Builtin') {
-      console.warn('Custom visualization module support is not yet implemented:', meta.module)
+  function toVisualizationCacheKey(meta: VisualizationIdentifier) {
+    return JSON.stringify({
+      // All fields MUST be explicitly written so that the order is consistent.
+      module: {
+        kind: meta.module.kind,
+        name: meta.module.kind === 'Library' ? meta.module.name : undefined,
+      },
+      name: meta.name,
+    }) as VisualizationCacheKey
+  }
+
+  function fromVisualizationCacheKey(key: VisualizationCacheKey): VisualizationIdentifier {
+    return JSON.parse(key)
+  }
+
+  async function onFileEvent(event: { path: Path; kind: FileEventKind }) {
+    const pathString = event.path.segments.join('/')
+    const name = currentProjectVisualizationsByPath.get(pathString)
+    const id: VisualizationIdentifier | undefined =
+      name != null ? { module: { kind: 'CurrentProject' }, name } : undefined
+    const key = id && toVisualizationCacheKey(id)
+    if (event.path.segments[0] === 'visualizations' && /\.vue$/.test(pathString)) {
+      compilationAbortControllers.get(pathString)?.abort()
+      compilationAbortControllers.delete(pathString)
+      switch (event.kind) {
+        case 'Added': {
+          try {
+            const abortController = new AbortController()
+            compilationAbortControllers.set(pathString, abortController)
+            const vizPromise = compile('enso-current-project:' + pathString)
+            const viz = await vizPromise
+            if (abortController.signal.aborted) break
+            const id: VisualizationIdentifier = {
+              module: { kind: 'CurrentProject' },
+              name: viz.name,
+            }
+            const key = toVisualizationCacheKey(id)
+            updateVisualizationTypes(id, viz.inputType)
+            cache.set(key, vizPromise)
+          } catch {
+            // Ignored - the file is either not a module, or contains an error.
```

Comment:
I think we should at least log the fact.

Reply:
for invalid formats i'm not sure what's the best thing to do -
there's a chance it's just a dependency, but there's also a chance it *is* just missing `name`

Reply:
Still, we should not silence them, even just for the sake of users complaining they don't see their visualizations in the list. Seeing the log, we immediately realize what's wrong.